### <font color='red'> Classification Task

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

data2: read the data for classification

In [3]:
data2= pd.read_csv("Twitter_S.csv")

check the information of data2:

In [4]:
data2.head()

0    1    2    3    4     5     6    7    8    9 ...   68   69   70   71  \
0  889  939  960  805  805  1143  1121  549  613  587 ...  1.0  1.0  889  939   
1  542  473  504  626  647   795   832  366  288  318 ...  1.0  1.0  542  473   
2   92   99  196  100  184    79   162   66   59  118 ...  1.0  1.0   92   99   
3   90   87   92  344  184   848   184   83   78   76 ...  1.0  1.0   90   87   
4  169   98  101   90   96    95   185  141   68   85 ...  1.0  1.0  169   98   

    72   73   74    75    76  77  
0  960  805  805  1143  1121   1  
1  504  626  647   795   832   1  
2  196  100  184    79   162   0  
3   92  344  184   848   184   1  
4  101   90   96    95   185   1  

[5 rows x 78 columns]

In [5]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140707 entries, 0 to 140706
Data columns (total 78 columns):
0     140707 non-null int64
1     140707 non-null int64
2     140707 non-null int64
3     140707 non-null int64
4     140707 non-null int64
5     140707 non-null int64
6     140707 non-null int64
7     140707 non-null int64
8     140707 non-null int64
9     140707 non-null int64
10    140707 non-null int64
11    140707 non-null int64
12    140707 non-null int64
13    140707 non-null int64
14    140707 non-null float64
15    140707 non-null float64
16    140707 non-null float64
17    140707 non-null float64
18    140707 non-null float64
19    140707 non-null float64
20    140707 non-null float64
21    140707 non-null float64
22    140707 non-null float64
23    140707 non-null float64
24    140707 non-null float64
25    140707 non-null float64
26    140707 non-null float64
27    140707 non-null float64
28    140707 non-null int64
29    140707 non-null int64
30    140707 non-null

In [6]:
data2.dtypes[0:5]

0    int64
1    int64
2    int64
3    int64
4    int64
dtype: object

In [7]:
data2.describe()

0              1              2              3  \
count  140707.000000  140707.000000  140707.000000  140707.000000   
mean      172.279823     155.150625     165.464476     176.820549   
std       509.872276     471.573236     495.360236     528.351277   
min         0.000000       0.000000       0.000000       0.000000   
25%         3.000000       2.000000       3.000000       3.000000   
50%        22.000000      19.000000      20.000000      22.000000   
75%       125.000000     112.000000     119.000000     126.000000   
max     24210.000000   22899.000000   20495.000000   27007.000000   

                   4              5              6              7  \
count  140707.000000  140707.000000  140707.000000  140707.000000   
mean      186.937700     216.209208     243.866510      87.050154   
std       560.331281     632.188378     707.402192     234.731748   
min         0.000000       0.000000       0.000000       0.000000   
25%         3.000000       4.000000       5.000000       2.000000   
50%        23.000000      28.000000      33.000000      13.000000   
75%       133.000000     161.000000     186.000000      70.000000   
max     30957.000000   28603.000000   37505.000000   15105.000000   

                   8              9      ...                   68  \
count  140707.000000  140707.000000      ...        140707.000000   
mean       78.639236      84.269574      ...             1.113444   
std       218.448179     233.536510      ...             1.374287   
min         0.000000       0.000000      ...             0.000000   
25%         2.000000       2.000000      ...             1.000000   
50%        11.000000      13.000000      ...             1.000000   
75%        64.000000      67.000000      ...             1.100000   
max     15730.000000   16389.000000      ...           185.666672   

                  69             70             71             72  \
count  140707.000000  140707.000000  140707.000000  140707.000000   
mean        1.196131     172.838807     155.630878     165.938674   
std         1.826150     510.937549     472.462733     496.233557   
min         0.000000       0.000000       0.000000       0.000000   
25%         1.000000       3.000000       2.000000       3.000000   
50%         1.000000      22.000000      19.000000      21.000000   
75%         1.119048     126.000000     113.000000     119.000000   
max       295.000000   24301.000000   22980.000000   20495.000000   

                  73             74             75             76  \
count  140707.000000  140707.000000  140707.000000  140707.000000   
mean      177.314810     187.463794     216.776294     244.479194   
std       529.286514     561.309487     633.203935     708.436795   
min         0.000000       0.000000       0.000000       0.000000   
25%         3.000000       3.000000       4.000000       6.000000   
50%        22.000000      23.000000      28.000000      33.000000   
75%       127.000000     134.000000     162.000000     187.000000   
max     27071.000000   31028.000000   28697.000000   37505.000000   

                  77  
count  140707.000000  
mean        0.197396  
std         0.398035  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 78 columns]

## common section for classification part

In [8]:
X=data2.iloc[:,0:77]
y=np.array(data2.iloc[:,77])

In [9]:
from sklearn.model_selection import train_test_split 
_,sample_data,_, sample_target = train_test_split(X, y, shuffle = True, test_size = 0.1)

## KNN classification

In [12]:
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import MinMaxScaler #standarized
from sklearn.model_selection import train_test_split #validation

X_train_org, X_test_org, y_train, y_test = train_test_split(sample_data,sample_target, random_state = 0)

scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train_org) #原本data frame變成array
X_test = scaler.transform(X_test_org)

knn = KNeighborsClassifier()
param_grid = {'n_neighbors':range(24,30)}

grid_knn = GridSearchCV(knn, param_grid=param_grid, cv = 5, scoring='roc_auc', n_jobs=-1) #n_jobs=-1:save some time
grid_knn.fit(X_train, y_train)
grid_knn.best_params_

{'n_neighbors': 27}

In [13]:
grid_knn.score(X_train, y_train)

0.9927258607737046

In [14]:
grid_knn.score(X_test, y_test)

0.9873221366923256

In [15]:
from sklearn.metrics import roc_auc_score
y_knn_predict = grid_knn.predict(X_test)
y_knn_train_predict = grid_knn.predict(X_train)
print('Train roc_auc_score: %.2f'%roc_auc_score(y_knn_train_predict, y_train))
print('Test roc_auc_score: %.2f '%roc_auc_score(y_knn_predict, y_test))

Train roc_auc_score: 0.95
Test roc_auc_score: 0.95 


In [99]:
report_table=[[]]
report_table = [['knn_classification', 'k = 27', grid_knn.score(X_train, y_train), grid_knn.score(X_test, y_test), roc_auc_score(y_knn_train_predict, y_train), roc_auc_score(y_knn_predict, y_test) ]]


## Logistic Regression

the first solution to find the best C for logistic regression by best score

In [18]:
from sklearn.pipeline import Pipeline
import sklearn.linear_model
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression()

pipe = Pipeline([('scaler', MinMaxScaler()), ('clf', LogisticRegression())])

create parameters for Pipeline and fit training data

In [19]:
param_gridl = {'clf__C': [0.001, 0.01, 0.1, 1, 10, 100, 1000], 'clf__penalty': ['l1','l2']}
grid_logit = GridSearchCV(pipe, param_grid=param_gridl,cv=5,scoring='roc_auc')
grid_logit.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('clf', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'clf__C': [0.001, 0.01, 0.1, 1, 10, 100, 1000], 'clf__penalty': ['l1', 'l2']},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

find the best scores and parameters

In [20]:
print("Best cross-validation accuracy: {:.2f}".format(grid_logit.best_score_))
print("Test set score: {:.2f}".format(grid_logit.score(X_test, y_test)))
print("Best parameters: {}".format(grid_logit.best_params_))

Best cross-validation accuracy: 0.99
Test set score: 0.99
Best parameters: {'clf__C': 1, 'clf__penalty': 'l1'}


In [21]:
y_logit_predict_train = grid_logit.predict(X_train)
y_logit_predict = grid_logit.predict(X_test)

In [100]:
report_table = report_table + [['logistic_regression', 'penalty=l1,C=1', grid_logit.score(X_train, y_train), grid_logit.score(X_test, y_test), roc_auc_score(y_logit_predict_train, y_train), roc_auc_score(y_logit_predict, y_test)]]

It seems that best C is 1 and penalty is l1

## Softmax Regression -> our dataset dependent variable is 0 and 1 so that we don't have to do the softmax 

## Build the pipeline for SVC with kernel trick

In [23]:
X_train = X_train_org
X_test = X_test_org

we don't have to do the MinMaxScaler because we will do that in Pipeline

## pipe line creation

## Pipe1: SVMlinear / Pipe2:SVMrbf / Pipe3:SVMpoly / Pipe4:LinearSVC

In [24]:
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC,SVC
from sklearn.model_selection import GridSearchCV


pipe1 = Pipeline([('scaler', MinMaxScaler()), ('svm', SVC(kernel = 'linear'))])

pipe2 = Pipeline([('scaler', MinMaxScaler()), ('svm', SVC(kernel = 'rbf'))])

pipe3 = Pipeline([('scaler', MinMaxScaler()), ('svm', SVC(kernel = 'poly'))])

pipe4 = Pipeline([('scaler', MinMaxScaler()), ('svm', LinearSVC())])

fit the training into each pipe variable

In [25]:
pipe1.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])

In [26]:
pipe2.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])

In [27]:
pipe3.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])

In [28]:
pipe4.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0))])

In [29]:
print("Test score: {:.2f}".format(pipe1.score(X_test, y_test)))
print("Test score: {:.2f}".format(pipe2.score(X_test, y_test)))
print("Test score: {:.2f}".format(pipe3.score(X_test, y_test)))
print("Test score: {:.2f}".format(pipe4.score(X_test, y_test)))

Test score: 0.96
Test score: 0.93
Test score: 0.82
Test score: 0.97


## parameters for Pipe1: SVMlinear / Pipe2:SVMrbf / Pipe3:SVMpoly / Pipe4: LinearSVC

In [91]:
param_grid1 = {'svm__C': [0.001, 0.01, 0.1, 1, 10, 100]}
param_grid2 = {'svm__C': [ 10,100,1000], 'svm__gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
param_grid3 = {'svm__C': [ 10,100,1000], 'svm__degree': [1, 2, 3]}
param_grid4 = {'svm__C': [0.001, 0.01, 0.1, 1, 10, 100]}

## search the best grid for each classification model

## SVMlinear

In [31]:
grid1 = GridSearchCV(pipe1, param_grid=param_grid1, cv=5, scoring='roc_auc')
grid1.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'svm__C': [0.001, 0.01, 0.1, 1, 10, 100]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [32]:
print("Best cross-validation accuracy: {:.2f}".format(grid1.best_score_))
print("Test set score: {:.2f}".format(grid1.score(X_test, y_test)))
print("Best parameters: {}".format(grid1.best_params_))

Best cross-validation accuracy: 0.99
Test set score: 0.99
Best parameters: {'svm__C': 10}


## SVMrbf

In [92]:
grid2 = GridSearchCV(pipe2, param_grid=param_grid2, cv=5, scoring='roc_auc')
grid2.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'svm__C': [10, 100, 1000], 'svm__gamma': [0.001, 0.01, 0.1, 1, 10, 100]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [93]:
print("Best cross-validation accuracy: {:.2f}".format(grid2.best_score_))
print("Test set score: {:.2f}".format(grid2.score(X_test, y_test)))
print("Best parameters: {}".format(grid2.best_params_))

Best cross-validation accuracy: 0.99
Test set score: 0.99
Best parameters: {'svm__C': 1000, 'svm__gamma': 0.01}


## SVMpoly

In [94]:
grid3 = GridSearchCV(pipe3, param_grid=param_grid3, cv=5, scoring='roc_auc')
grid3.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'svm__C': [10, 100, 1000], 'svm__degree': [1, 2, 3]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [95]:
print("Best cross-validation accuracy: {:.2f}".format(grid3.best_score_))
print("Test set score: {:.2f}".format(grid3.score(X_test, y_test)))
print("Best parameters: {}".format(grid3.best_params_))

Best cross-validation accuracy: 0.99
Test set score: 0.99
Best parameters: {'svm__C': 1000, 'svm__degree': 1}


## LinearSVC

In [37]:
grid4 = GridSearchCV(pipe4, param_grid=param_grid4, cv=5, scoring='roc_auc')
grid4.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('scaler', MinMaxScaler(copy=True, feature_range=(0, 1))), ('svm', LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'svm__C': [0.001, 0.01, 0.1, 1, 10, 100]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [38]:
print("Best cross-validation accuracy: {:.2f}".format(grid4.best_score_))
print("Test set score: {:.2f}".format(grid4.score(X_test, y_test)))
print("Best parameters: {}".format(grid4.best_params_))

Best cross-validation accuracy: 0.99
Test set score: 0.99
Best parameters: {'svm__C': 10}


In [101]:
y_SVMlinear_predict = grid1.predict(X_test)
y_SVMlinear_train_predict = grid1.predict(X_train)
y_SVMrbf_predict = grid2.predict(X_test)
y_SVMrbr_train_predict = grid2.predict(X_train)
y_SVMpoly_predict = grid3.predict(X_test)
y_SVMpoly_train_predict = grid3.predict(X_train)
y_LinearSVC_predict = grid4.predict(X_test)
y_LinearSVC_train_predict = grid4.predict(X_train)
report_table = report_table+[['SVMlinear', 'C = 10', grid1.score(X_train, y_train), grid1.score(X_test, y_test), roc_auc_score(y_SVMlinear_train_predict, y_train), roc_auc_score(y_SVMlinear_predict, y_test) ]]
report_table = report_table+[['SVMrbf', 'gamma=0.1, C=1000', grid2.score(X_train, y_train), grid2.score(X_test, y_test), roc_auc_score(y_SVMrbr_train_predict, y_train), roc_auc_score(y_SVMrbf_predict , y_test) ]]
report_table = report_table+[['SVMpoly', 'degree=1, C=1000', grid3.score(X_train, y_train), grid3.score(X_test, y_test), roc_auc_score(y_SVMpoly_train_predict, y_train), roc_auc_score(y_SVMpoly_predict, y_test) ]]
report_table = report_table+[['LinearSVC', 'C = 10', grid4.score(X_train, y_train), grid4.score(X_test, y_test), roc_auc_score(y_LinearSVC_train_predict, y_train), roc_auc_score(y_LinearSVC_predict, y_test) ]]


## Decision Tree

In [40]:
%matplotlib notebook
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

In [41]:
dtree = DecisionTreeClassifier(random_state=0)

dtree.fit(X_train, y_train)
print("Accuracy on training set: {:.3f}".format(dtree.score(X_train, y_train)))
print("Accuracy on test set: {:.3f}".format(dtree.score(X_test, y_test)))

Accuracy on training set: 1.000
Accuracy on test set: 0.953


In [42]:
dtree = DecisionTreeClassifier(max_depth=4, random_state=0)  
dtree.fit(X_train, y_train)

print("Accuracy on training set: {:.3f}".format(dtree.score(X_train, y_train)))
print("Accuracy on test set: {:.3f}".format(dtree.score(X_test, y_test)))

Accuracy on training set: 0.969
Accuracy on test set: 0.961


citation: https://machinelearningmastery.com/feature-importance-and-feature-selection-with-xgboost-in-python/?fbclid=IwAR1Wg1-QXqJRVmC6Jd3moc1W8FxapS3WQilJpwedj0fdtuU7l-5bdtypq2g

This importance is calculated explicitly for each attribute in the dataset, allowing attributes to be ranked and compared to each other.

In [102]:
dtree = DecisionTreeClassifier()
dtree.fit(X_train, y_train)
y_dtree_predict = dtree.predict(X_test)
y_dtree_train_predict = dtree.predict(X_train)
report_table = report_table+[['decision_tree', 'none', dtree.score(X_train, y_train), dtree.score(X_test, y_test), roc_auc_score(y_dtree_train_predict, y_train), roc_auc_score(y_dtree_predict, y_test) ]]

In [103]:
report = pd.DataFrame(report_table,columns = ['Model name', 'Model parameter', 'Train accuracy', 'Test accuracy', 'Train auc score', 'Test auc score'])
report.index = report['Model name']
report

Model name    Model parameter  Train accuracy  \
Model name                                                                    
knn_classification    knn_classification             k = 27        0.533550   
logistic_regression  logistic_regression     penalty=l1,C=1        0.991447   
SVMlinear                      SVMlinear             C = 10        0.990984   
SVMrbf                            SVMrbf  gamma=0.1, C=1000        0.991421   
SVMpoly                          SVMpoly   degree=1, C=1000        0.991117   
LinearSVC                      LinearSVC             C = 10        0.991591   
decision_tree              decision_tree               none        1.000000   

                     Test accuracy  Train auc score  Test auc score  
Model name                                                           
knn_classification        0.529741         0.950001        0.953181  
logistic_regression       0.992841         0.957205        0.960973  
SVMlinear                 0.992494         0.955494        0.961561  
SVMrbf                    0.992470         0.955711        0.960450  
SVMpoly                   0.992508         0.955494        0.959475  
LinearSVC                 0.992766         0.957443        0.961498  
decision_tree             0.953667         1.000000        0.922997

### Summary: 

Due to large dataset, we pick up 10% sample data for modeling. On knn model, We get 0.95 for auc training score and 0.953 for auc testing score. For logiatic model, the auc training score is 0.957, and the auc testing score is 0.961. On SVM model, the auc training score is 0.955 and the auc testing score is 0.962 for SVMlinear. The auc training score is 0.956 and the auc testing score is 0.96 for SVMrbf. The auc training score is 0.955 and the auc testing score is 0.959 for SVMpoly. The auc training score is 0.957 and the auc testing score is 0.961 for LinearSVC. And the followings are score for decision tree: auc training score: 1.000, auc testing score: 0.923.

Accoring to the result, we choose the model which gets highest auc testing score.That is SVM_linear model for classfication.

### Run whole data on LinearSVC model:

In [104]:
X_train_org, X_test_org, y_train, y_test = train_test_split(X,y, random_state = 0)
X_train = X_train_org
X_test = X_test_org

In [105]:
model=LinearSVC(C=10)
model.fit(X_train, y_train)

LinearSVC(C=10, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)

In [106]:
model.coef_

array([[ 4.08822856e-03, -6.45495648e-03,  1.94492130e-02,
         6.79787343e-04, -5.21592853e-03, -2.36887538e-03,
        -2.74686680e-03, -1.22085376e-03, -1.58645655e-03,
         7.26126122e-05,  1.45485927e-03,  1.26071774e-04,
         1.65667339e-03,  4.74703174e-05,  2.71426622e-05,
         8.16057877e-06,  3.88546190e-05,  4.50554823e-05,
         2.23805877e-05,  2.55365133e-05, -8.17358649e-05,
         1.58769585e-02,  4.09128653e-03,  1.32514180e-02,
         1.79791338e-03, -5.51583442e-03, -8.65117691e-02,
         1.67474545e-02,  1.89606278e-03,  3.02750080e-04,
         1.31319354e-03,  3.91737447e-04,  1.86043735e-04,
        -1.94104687e-03,  8.29919115e-04,  2.38034792e-05,
         8.42374238e-06,  2.71172311e-05,  3.14146306e-05,
         1.47965551e-05,  1.46302728e-05, -4.03799944e-05,
        -8.75954059e-03,  1.12769143e-02, -8.39451943e-03,
         3.35491389e-03,  1.93264685e-02, -8.78722567e-02,
         1.46477523e-02, -5.43217656e-03, -2.55178655e-0

In [107]:
model.intercept_

array([-1.02109152])

In [109]:
print("training score:",model.score(X_train,y_train))
print("testing score:",model.score(X_test,y_test))

training score: 0.8737041599545153
testing score: 0.8730420445177246
